File: scanner.go
package main

import (
	"os/exec"
	"strings"
)

type Action struct {
	Label   string
	Command string
}

type Scanner struct {
	Actions  []Action
	Detected []string
	Status   string
}

func NewScanner() *Scanner {
	return &Scanner{
		Actions:  []Action{},
		Detected: []string{},
		Status:   "Ready.",
	}
}

func (s *Scanner) ScanActions() {
	s.Actions = nil
	s.Detected = nil

	if !shSuccess("command -v pkexec >/dev/null 2>&1") {
		s.Status = "pkexec not found. Install polkit first."
		return
	}

	grub := ""
	if shSuccess("command -v grub-mkconfig >/dev/null 2>&1") {
		grub = "grub-mkconfig"
	} else if shSuccess("command -v grub2-mkconfig >/dev/null 2>&1") {
		grub = "grub2-mkconfig"
	}

	if grub != "" {
		if shSuccess("[ -e '/boot/grub' ]") {
			s.addAction("Update GRUB", buildPkexecCommand(grub, []string{"-o", "/boot/grub/grub.cfg"}), "GRUB")
		} else if shSuccess("[ -e '/boot/grub2' ]") {
			s.addAction("Update GRUB", buildPkexecCommand(grub, []string{"-o", "/boot/grub2/grub.cfg"}), "GRUB")
		}
	}

	if shSuccess("command -v bootctl >/dev/null 2>&1") &&
		(shSuccess("[ -e '/boot/loader' ]") || shSuccess("[ -e '/efi/loader' ]")) {
		s.addAction("Update systemd-boot", buildPkexecCommand("bootctl", []string{"update"}), "systemd-boot")
	}

	if shSuccess("command -v refind-install >/dev/null 2>&1") {
		s.addAction("Refresh rEFInd", buildPkexecCommand("refind-install", []string{}), "rEFInd")
	}

	if shSuccess("command -v mkinitcpio >/dev/null 2>&1") {
		s.addAction("Rebuild initramfs", buildPkexecCommand("mkinitcpio", []string{"-P"}), "mkinitcpio")
	} else if shSuccess("command -v dracut >/dev/null 2>&1") {
		s.addAction("Rebuild initramfs", buildPkexecCommand("dracut", []string{"--regenerate-all", "--force"}), "dracut")
	} else if shSuccess("command -v update-initramfs >/dev/null 2>&1") {
		s.addAction("Rebuild initramfs", buildPkexecCommand("update-initramfs", []string{"-u", "-k", "all"}), "update-initramfs")
	}

	if len(s.Actions) > 0 {
		s.Status = "Detected: " + strings.Join(s.Detected, ", ")
	} else {
		s.Status = "Nothing supported detected."
	}
}

func (s *Scanner) addAction(label string, command string, detectName string) {
	s.Actions = append(s.Actions, Action{
		Label:   label,
		Command: command,
	})
	s.Detected = append(s.Detected, detectName)
}

func shSuccess(command string) bool {
	cmd := exec.Command("sh", "-c", command)
	err := cmd.Run()
	if err != nil {
		return false
	}
	return cmd.ProcessState.Success()
}

func quoteSinglePosix(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\\''") + "'"
}

func buildPkexecCommand(binary string, args []string) string {
	cmd := "pkexec " + quoteSinglePosix(binary)
	for _, a := range args {
		cmd += " " + quoteSinglePosix(a)
	}
	return cmd
}

File: src/scanner.rs
use std::process::Command;

#[derive(Clone, Debug)]
pub struct Action {
    pub label: String,
    pub command: String,
}

pub struct Scanner {
    pub actions: Vec<Action>,
    pub detected: Vec<String>,
    pub status: String,
}

impl Scanner {
    pub fn new() -> Self {
        Self {
            actions: Vec::new(),
            detected: Vec::new(),
            status: "Ready.".to_string(),
        }
    }

    pub fn scan_actions(&mut self) {
        self.actions.clear();
        self.detected.clear();

        if !sh_success("command -v pkexec >/dev/null 2>&1") {
            self.status = "pkexec not found. Install polkit first.".to_string();
            return;
        }

        let grub = if sh_success("command -v grub-mkconfig >/dev/null 2>&1") {
            Some("grub-mkconfig")
        } else if sh_success("command -v grub2-mkconfig >/dev/null 2>&1") {
            Some("grub2-mkconfig")
        } else {
            None
        };

        if let Some(grub) = grub {
            if sh_success("[ -e '/boot/grub' ]") {
                self.add_action(
                    "Update GRUB",
                    &build_pkexec_command(grub, &["-o", "/boot/grub/grub.cfg"]),
                    "GRUB",
                );
            } else if sh_success("[ -e '/boot/grub2' ]") {
                self.add_action(
                    "Update GRUB",
                    &build_pkexec_command(grub, &["-o", "/boot/grub2/grub.cfg"]),
                    "GRUB",
                );
            }
        }

        if sh_success("command -v bootctl >/dev/null 2>&1")
            && (sh_success("[ -e '/boot/loader' ]") || sh_success("[ -e '/efi/loader' ]"))
        {
            self.add_action(
                "Update systemd-boot",
                &build_pkexec_command("bootctl", &["update"]),
                "systemd-boot",
            );
        }

        if sh_success("command -v refind-install >/dev/null 2>&1") {
            self.add_action(
                "Refresh rEFInd",
                &build_pkexec_command("refind-install", &[]),
                "rEFInd",
            );
        }

        if sh_success("command -v mkinitcpio >/dev/null 2>&1") {
            self.add_action(
                "Rebuild initramfs",
                &build_pkexec_command("mkinitcpio", &["-P"]),
                "mkinitcpio",
            );
        } else if sh_success("command -v dracut >/dev/null 2>&1") {
            self.add_action(
                "Rebuild initramfs",
                &build_pkexec_command("dracut", &["--regenerate-all", "--force"]),
                "dracut",
            );
        } else if sh_success("command -v update-initramfs >/dev/null 2>&1") {
            self.add_action(
                "Rebuild initramfs",
                &build_pkexec_command("update-initramfs", &["-u", "-k", "all"]),
                "update-initramfs",
            );
        }

        self.status = if !self.actions.is_empty() {
            format!("Detected: {}", self.detected.join(", "))
        } else {
            "Nothing supported detected.".to_string()
        };
    }

    fn add_action(&mut self, label: &str, command: &str, detect_name: &str) {
        self.actions.push(Action {
            label: label.to_string(),
            command: command.to_string(),
        });
        self.detected.push(detect_name.to_string());
    }
}

fn sh_success(command: &str) -> bool {
    // We use /bin/sh via shell -c so redirects/[] checks work.
    Command::new("sh")
        .arg("-c")
        .arg(command)
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

fn quote_single_posix(s: &str) -> String {
    // safe quoting for POSIX sh single quotes
    // 'foo' => '\'' pattern
    format!("'{}'", s.replace("'", "'\\''"))
}

fn build_pkexec_command(binary: &str, args: &[&str]) -> String {
    let mut cmd = format!("pkexec {}", quote_single_posix(binary));
    for a in args {
        cmd.push(' ');
        cmd.push_str(&quote_single_posix(a));
    }
    cmd
}
